pilot/pkg/model/config.go

```diff
@@ -277,48 +277,46 @@ func resolveGatewayName(gwname string, meta config.Meta) string {
 	return out
 }
 
-// MostSpecificHostMatch compares the map of the stack to the needle, and returns the longest element
-// matching the needle, or false if no element in the map matches the needle.
-func MostSpecificHostMatch[V any](needle host.Name, m map[host.Name]V) (host.Name, bool) {
-	matches := []host.Name{}
-	// exact match first
-	if m != nil {
-		if _, ok := m[needle]; ok {
-			return needle, true
+// MostSpecificHostMatch compares the maps of specific and wildcard hosts to the needle, and returns the longest element
+// matching the needle and it's value, or false if no element in the maps matches the needle.
+func MostSpecificHostMatch[V any](needle host.Name, specific map[host.Name]V, wildcard map[host.Name]V) (host.Name, V, bool) {
+	if needle.IsWildCarded() {
+		// exact match first
+		if v, ok := wildcard[needle]; ok {
+			return needle, v, true
 		}
+
+		return mostSpecificHostWildcardMatch(string(needle[1:]), wildcard)
 	}
 
-	if needle.IsWildCarded() {
-		for h := range m {
-			// both needle and h are wildcards
-			if h.IsWildCarded() {
-				if len(needle) < len(h) {
-					continue
-				}
-				if strings.HasSuffix(string(needle[1:]), string(h[1:])) {
-					matches = append(matches, h)
-				}
-			}
-		}
-	} else {
-		for h := range m {
-			// only n is wildcard
-			if h.IsWildCarded() {
-				if strings.HasSuffix(string(needle), string(h[1:])) {
-					matches = append(matches, h)
-				}
+	// exact match first
+	if v, ok := specific[needle]; ok {
+		return needle, v, true
+	}
+
+	// check wildcard
+	return mostSpecificHostWildcardMatch(string(needle), wildcard)
+}
+
+func mostSpecificHostWildcardMatch[V any](needle string, wildcard map[host.Name]V) (host.Name, V, bool) {
+	found := false
+	var matchHost host.Name
+	var matchValue V
+
+	for h, v := range wildcard {
+		if strings.HasSuffix(needle, string(h[1:])) {
+			if !found {
+				matchHost = h
+				matchValue = wildcard[h]
+				found = true
+			} else if host.MoreSpecific(h, matchHost) {
```

Comment:
Not sure what is MoreSpecific mean?

Reply:
why not use original check

Reply:
It is the same check as before, just a different name.
Before we sorted the array of matches, sorting used method `Less`, I just moved that to a new method called `MoreSpecific` to be a little more clear.

Reply:
The method also has a comment saying what it does

Reply:
We previously used `sort.Sort()` which takes a `sort.Interface` parameter which is the array to sort, `host.Names` implements that interface, one of it's methods is `Less(int i, j) bool` which returns if one position should be sorted previously than the other. So basically, `host.Names.Less()` was dictating which hostnames are more specific (should be sorted first). I extracted that logic to a separate method to be able to use it directly, and make `Less()` call this method.